changelog/dmd.isCOMClass.dd

```diff
@@ -0,0 +1,5 @@
+New trait isCOMClass to detect if a type is a COM class
+
+A COM class inherits from a possibly user defined interface called ``IUnknown``.
+To detect this during compilation use the trait ``__traits(isCOMClass, Type)``.
+Or using the ``TypeInfo_Class`` flag.
```

Comment:
How do you check that flag? If this works as an alternative, why is `__traits(isCOMClass)` introduced?

Reply:
Here is the original code from the hook:

```d
extern (C) Object _d_newclass(const ClassInfo ci) @weak
{
    import core.stdc.stdlib;
    import core.exception : onOutOfMemoryError;
    void* p;
    auto init = ci.initializer;

    debug(PRINTF) printf("_d_newclass(ci = %p, %s)\n", ci, cast(char *)ci.name);
    if (ci.m_flags & TypeInfo_Class.ClassFlags.isCOMclass)
 ```
 
 The templated hook that has been fixed to use the new trait:
 
 ```d
T _d_newclassT(T)() @trusted
if (is(T == class))
{
    import core.internal.traits : hasIndirections;
    import core.exception : onOutOfMemoryError;
    import core.memory : pureMalloc;
    import core.memory : GC;

    alias BlkAttr = GC.BlkAttr;

    auto init = __traits(initSymbol, T);
    void* p;

    static if (__traits(isCOMClass, T))
```

While both do the same thing new a class, one uses purely CT available information, the other RT only information.

Reply:
Either we revert back to TypeInfo for this hook, or we go in the direction of some sort of CT available information.

Of course we could derive it by looking for a ``IUnknown`` interface, but that isn't an easy thing to do, let alone perform on every single class new instantiation when there is a flag in the compiler easily readable with the information we need.

Reply:
The new trait is good, I just think the last line of the changelog should be removed or clarified. The current wording implies `TypeInfo_Class` can be used during compilation to check the same thing.

Reply:
Done.

Reply:
Something like "The traits is equivalent to the runtime check `ClassInfo.m_flags & TypeInfo_Class.ClassFlags.isCOMclass` ".

Reply:
I don't want to encourage people to use anything TypeInfo-related, it's stated so that people know what existing code they can replace with the trait.